File: benchmarks/cross-language/go/thread_ring.go
// Go Thread Ring Benchmark (Savina-style)
// Uses buffered channels to model actor mailboxes
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func getMessages() int {
	if env := os.Getenv("BENCHMARK_MESSAGES"); env != "" {
		if n, err := strconv.Atoi(env); err == nil {
			return n
		}
	}
	return 100000
}

// Match other implementations: 100 nodes, configurable hops
const ringSize = 100

var numHops = getMessages()

func ringNode(id int, inbox <-chan int, next chan<- int, done chan<- int) {
	received := 0
	for token := range inbox {
		received++
		if token == 0 {
			done <- received
			return
		}
		next <- token - 1
	}
}

func main() {
	fmt.Println("=== Go Thread Ring Benchmark ===")
	fmt.Printf("Ring size: %d, Hops: %d\n\n", ringSize, numHops)

	// Buffered channels model actor mailboxes
	// Size 1 maintains ordering while allowing async message passing
	channels := make([]chan int, ringSize)
	for i := 0; i < ringSize; i++ {
		channels[i] = make(chan int, 1)
	}

	done := make(chan int)

	// Start goroutines
	for i := 0; i < ringSize; i++ {
		next := channels[(i+1)%ringSize]
		go ringNode(i, channels[i], next, done)
	}

	start := time.Now()

	// Inject initial token
	channels[0] <- numHops

	// Wait for completion
	received := <-done

	elapsed := time.Since(start)

	totalMessages := numHops + 1
	// Each node receives totalMessages/ringSize messages
	expectedPerNode := totalMessages/ringSize + 1
	if received < expectedPerNode-10 || received > expectedPerNode+10 {
		fmt.Printf("VALIDATION WARNING: expected ~%d per node, got %d\n", expectedPerNode, received)
	}

	elapsedSec := elapsed.Seconds()
	throughput := float64(totalMessages) / elapsedSec / 1e6
	nsPerMsg := elapsedSec * 1e9 / float64(totalMessages)

	fmt.Printf("ns/msg:         %.2f\n", nsPerMsg)
	fmt.Printf("Throughput:     %.2f M msg/sec\n", throughput)
}

File: benchmarks/cross-language/rust/src/thread_ring.rs
// Rust Thread Ring Benchmark (Savina-style)
use std::sync::mpsc::{channel, Sender, Receiver};
use std::thread;
use std::time::Instant;

const RING_SIZE: usize = 100;

fn get_messages() -> usize {
    std::env::var("BENCHMARK_MESSAGES")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100_000)
}

fn ring_node(rx: Receiver<usize>, next_tx: Sender<usize>, done_tx: Sender<usize>) {
    let mut received = 0;
    while let Ok(token) = rx.recv() {
        received += 1;
        if token == 0 {
            done_tx.send(received).unwrap();
            return;
        }
        next_tx.send(token - 1).unwrap();
    }
}

fn main() {
    let num_hops = get_messages();
    println!("=== Rust Thread Ring Benchmark ===");
    println!("Ring size: {}, Hops: {}\n", RING_SIZE, num_hops);

    let (done_tx, done_rx) = channel();

    // Create channels for the ring
    let mut senders: Vec<Sender<usize>> = Vec::with_capacity(RING_SIZE);
    let mut receivers: Vec<Receiver<usize>> = Vec::with_capacity(RING_SIZE);

    for _ in 0..RING_SIZE {
        let (tx, rx) = channel();
        senders.push(tx);
        receivers.push(rx);
    }

    // Start threads
    let mut handles = Vec::with_capacity(RING_SIZE);
    for i in 0..RING_SIZE {
        let rx = receivers.remove(0);
        let next_tx = senders[(i + 1) % RING_SIZE].clone();
        let done = done_tx.clone();
        handles.push(thread::spawn(move || ring_node(rx, next_tx, done)));
    }
    drop(done_tx);

    let start = Instant::now();

    // Inject token
    senders[0].send(num_hops).unwrap();

    // Wait for completion
    let received = done_rx.recv().unwrap();

    let elapsed = start.elapsed().as_secs_f64();

    // Cleanup - drop senders to unblock threads
    drop(senders);
    for h in handles {
        let _ = h.join();
    }

    let total_messages = num_hops + 1;
    // Each node receives total_messages / ring_size messages
    let expected_per_node = total_messages / RING_SIZE + 1;
    if received < expected_per_node - 10 || received > expected_per_node + 10 {
        println!("VALIDATION WARNING: expected ~{} per node, got {}", expected_per_node, received);
    }

    let throughput = total_messages as f64 / elapsed / 1e6;
    let ns_per_msg = elapsed * 1e9 / total_messages as f64;

    println!("ns/msg:         {:.2}", ns_per_msg);
    println!("Throughput:     {:.2} M msg/sec", throughput);
}
